# HG changeset 9689 patch
From: kaf24@firebug.cl.cam.ac.uk
# Node ID 5f2d82cdbb894ad3c334a844b2d9b6c308189c0b
# Parent  f75e01cfde7613b0fcd0a5dce41765df75af044f
Subject Remove unnecessary x86-specific mapping code from privcmd interface.
Signed-off-by: Keir Fraser <keir@xensource.com>
xen-unstable changeset:   10150:2436cc71de508f409b957e9eb24c1f34b1ab615d
xen-unstable date:        Wed May 24 14:49:24 2006 +0100

Acked-by: Jan Beulich <jbeulich@novell.com>

Index: head-2006-05-26/drivers/xen/privcmd/privcmd.c
===================================================================
--- head-2006-05-26.orig/drivers/xen/privcmd/privcmd.c	2006-05-26 11:54:16.000000000 +0200
+++ head-2006-05-26/drivers/xen/privcmd/privcmd.c	2006-05-26 11:54:25.000000000 +0200
@@ -159,12 +159,10 @@ static int privcmd_ioctl(struct inode *i
 	break;
 
 	case IOCTL_PRIVCMD_MMAPBATCH: {
-		mmu_update_t u;
 		privcmd_mmapbatch_t m;
 		struct vm_area_struct *vma = NULL;
 		unsigned long __user *p;
 		unsigned long addr, mfn; 
-		uint64_t ptep;
 		int i;
 
 		if (copy_from_user(&m, udata, sizeof(m))) {
@@ -207,15 +205,10 @@ static int privcmd_ioctl(struct inode *i
 			if (ret < 0)
 			    goto batch_err;
 #else
-
-			ret = create_lookup_pte_addr(vma->vm_mm, addr, &ptep);
-			if (ret)
-				goto batch_err;
-
-			u.val = pte_val_ma(pfn_pte_ma(mfn, vma->vm_page_prot));
-			u.ptr = ptep;
-
-			if (HYPERVISOR_mmu_update(&u, 1, NULL, m.dom) < 0)
+			ret = direct_remap_pfn_range(vma, addr & PAGE_MASK,
+						     mfn, PAGE_SIZE,
+						     vma->vm_page_prot, m.dom);
+			if (ret < 0)
 				put_user(0xF0000000 | mfn, p);
 #endif
 		}
